### ISSUE TITLE
Virtual Component Groups in the Hierarchical Action List

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4381]
This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/181865548).
<hr />

## Why

**As** Enso IDE developer
**I want** to have access to Virtual Component Groups in new Hierarchical Action List
**So that** I can implement displaying them in Component List Panel View.

## Acceptance Criteria

```gherkin
Scenario: 
GIVEN I have test with mocked Engine's response for `executionContext/getComponentGroups`
WHEN I create searcher controller without this argument and with no input
THEN I receive proper Virtual Groups in Action List in proper order
```

## Implementation Notes

 - #182304840 was created to fix the order of component groups returned in the response to the `executionContext/getComponentGroups` message.
 - The list of component groups retrieved from the Engine should be stored in the `ExecutionContext`. The list should contain qualified names of entries that will allow retrieving suggestion entry IDs from the Suggestion Database.
 - Refreshing the component groups list when the `package.yaml` file changes is not in scope of this task.

## Tasks:
- [X] Parse Engine's response containing Virtual Component Groups and store the results in the Execution Context model.
- [ ] Put information about Virtual Component Groups in the Hierarchical Actions List (needs #181855501)

## Comments:
To clarify: the "in proper order" fragment in the ACs means the order defined in the [Component Virtual Groups ordering](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#component-virtual-groups-ordering) section of the Design Doc, is that right? Thus, we'll need to mock multiple Engine responses, one for each imported library, yes?

Also, the "Why" section of this Task's Description mentions a "Hierarchical Action List" - does this mean that this Task is blocked by #181855501? (Mateusz Czapliński - Apr 14, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-05-25):

**Progress:** Explored the Task description, Design Doc, and Language Server protocol, to understand the Task scope. Discussed Tech Design with AO. Asked DB a few questions crucial to the Tech Design of the task on #engine. Started working on retrieving a sample response to an `executionContext/getComponentGroups` message from Engine. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Check answers from Dmitry and continue discussing the questions if needed. Add an initial test parsing a sample  `executionContext/getComponentGroups` response from Engine (base the test on similar existing tests). Start working on adding a database of virtual component groups in the Execution Context model struct. Update the Task in Pivotal with details after the Tech Design talk. Also, merge #181852566 if Acc/QA Review is Passed. (Enso Bot - May 26, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-05-26):

**Progress:** Got confirmation that a response to the `executionContext/getComponentGroups` message is expected to contain local entries and that it is expected to return results ordered as in the Design Doc. The order is currently broken and a ticket was opened to fix it (#182304840). Wrote a PoC and test for the subtask 1/2 in the current Ticket. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Improve and cleanup the code for a PR of subtask 1/2. Open a PR. Start working on the subtask 2/2. (Enso Bot - May 27, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for the provided date (2022-05-27):

**Progress:** Improved the initial draft code for subtask 1/2 and prepared it for a PR. Opened a PR. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Discuss with AO how to proceed with subtask 2/2 given that the code currently published in a branch for #181855501 (a dependency of the subtask) appears old and not useful. A standup for #181855501 claims a more advanced stage of the code, so hopefully a push to the branch by AO might unblock me. (Enso Bot - May 30, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-05-30):

**Progress:** Addressed comments from 1st code-review by AO, re-requested code-review. Read the PR by AO in preparation for starting to work on subtask 2/2. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Address code-review comments by WD. If time permits, start implementing subtask 2/2. (Enso Bot - May 31, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-05-31):

**Progress:** Addressed some comments from code-review by WD and asked him on Discord for advice and clarification on others. Did first draft of code for subtask 2/2 based on AO's branch. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Hopefully get answers from WD and thus be able to address the remaining code-review comments. Subtask 2/2: add color handling; ask AO about status of his branch; start a cleanup of the code and consider refactorings in preparation for opening a PR. (Enso Bot - Jun 1, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-01):

**Progress:** Addressed code-review comments for subtask 1/2 and discussed a number of them with WD and AO. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Address remaining code-review comments for subtask 1/2. Continue working on subtask 2/2. (Enso Bot - Jun 2, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-02):

**Progress:** Addressed comments for next code-review for subtask 1/2. Requested Acc/QA Review for subtask 1. Did some work on subtask 2/2. It should be finished by 2022-06-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Do Acc/QA Review round 2 for #181855501. Then get back to subtask 2/2, polish it for a PR. (Enso Bot - Jun 3, 2022)
<hr />
**Mateusz Czapliński** reports a new **🔴 DELAY** for today (2022-06-03):

**Summary:** There is **1 day** delay in implementation of the [Virtual Component Groups in the Hierarchical Action List (#181865548) task](https://www.pivotaltracker.com/story/show/181865548).
It **will cause 1 day delay** for the delivery of this weekly plan.

**Delay Cause:** I spent 1 day doing two rounds of Acc/QA Review on #181855501. The subtask 2/2 of #181865548 I'm working on is also depending on #181855501 and I cannot open a PR for the subtask until #181855501 is merged, and until subtask 1/2 is merged (PR 3488). Both mentioned dependency PRs now have "CI: Ready to merge" label applied, so they should get merged over then weekend, and I should be able to open a PR of subtask 2/2 on Monday. (Enso Bot - Jun 3, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for the provided date (2022-06-03):

**Progress:** Merged #181855501 and updated the code to adjust to changes in the merged branch. Started polishing and self-reviewing for a PR. It should be finished by 2022-06-07.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Discuss some doubts with AO. Finish polishing the code, perform some integration tests and open a PR. (Enso Bot - Jun 6, 2022)
<hr />
**Mateusz Czapliński** reports a new **🔴 DELAY** for today (2022-06-06):

**Summary:** There is **1 day** delay in implementation of the [Virtual Component Groups in the Hierarchical Action List (#181865548) task](https://www.pivotaltracker.com/story/show/181865548).
It **will cause 1 day delay** for the delivery of this weekly plan.

**Delay Cause:** After a discussion with AO, a decision was made to implement a small optimization in the design by caching virtual component groups in the Node Searcher. (Enso Bot - Jun 6, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-06):

**Progress:** Discussed an optimization with AO and implemented it. It should be finished by 2022-06-08.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Add some debugging messages and test with Engine to verify that the code actually works correctly not only in tests. Finish polishing the code and open a PR. (Enso Bot - Jun 7, 2022)
<hr />
**Mateusz Czapliński** reports a new **🔴 DELAY** for today (2022-06-07):

**Summary:** There is **1 day** delay in implementation of the [Virtual Component Groups in the Hierarchical Action List (#181865548) task](https://www.pivotaltracker.com/story/show/181865548).
It **will cause 1 day delay** for the delivery of this weekly plan.

**Delay Cause:** I did a manual integration test with the Engine. Unexpectedly (the behavior doesn't seem documented and wasn't anticipated during the task design), I discovered that Engine responds with empty component groups immediately after creation of an execution context. Further experimenting showed that component groups seem properly filled after an `executionCompleted` message is received from the Engine (I also asked a question on #engine but didn't get any answer today). This resulted in a need for non-trivial modifications in the code and tests. At the current moment, the code & tests I have are updated, but again I still need to do some polishing of the result before opening a PR. Thus, I estimate to open a PR only tomorrow. (Enso Bot - Jun 7, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-07):

**Progress:** Run manual integration test with Engine and noticed empty response to `getComponentGroups` request. Found out that the response is non-empty if requested after `executionCompleted` is received by the IDE. Changed the code and tests to send `getComponentGroups` after `executionCompleted` is received. It should be finished by 2022-06-09.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Polish the code, self-review and open a PR. (Enso Bot - Jun 8, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-08):

**Progress:** Did some final polishing and opened a PR. It should be finished by 2022-06-09. (Enso Bot - Jun 9, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for yesterday (2022-06-09):

**Progress:** Addressed code-review comments and re-requested review. It should be finished by 2022-06-09.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181865548). Address code-review comments if new ones show up. Otherwise, work on #181869186. (Enso Bot - Jun 10, 2022)
<hr />
**Mateusz Czapliński** reports a new **🔴 DELAY** for today (2022-06-10):

**Summary:** There is **4 days** delay in implementation of the [Virtual Component Groups in the Hierarchical Action List (#181865548) task](https://www.pivotaltracker.com/story/show/181865548).
It **will cause 3 days delay** for the delivery of this weekly plan.

2 days of weekend + 1 day code-review + 1 day Acc/QA review (not included in Weekly Plan Delay). In meantime, I've been working when possible on #181869186, which also mitigates weekly plan delay.

**Delay Cause:** Initial code-review had a delay + there's a 2nd round of code-review. (Enso Bot - Jun 10, 2022)
<hr />
**Mateusz Czapliński** reports a new **STANDUP** for the provided date (2022-06-10):

**Progress:** Addressed most code review comments and asked WD about a few not fully clear to me. It should be finished by 2022-06-13. (Enso Bot - Jun 13, 2022)
<hr />
**Acceptance/QA** review set to **pass** (Adam Obuchowicz - Jun 15, 2022)
<hr />
